ironfish/src/network/transactionFetcher.ts

```diff
@@ -0,0 +1,181 @@
+
+    const sources = this.sources.get(hash) || new Set<Peer>()
+    sources.add(peer)
+    this.sources.set(hash, sources)
+
+    if (!currentState) {
+      this.fetchFromNextSource(hash)
+    }
+  }
+
+  /* Called when a transaction is received and confirmed from the network
+   * either in a block or in a gossiped transaction request */
+  resolve(transaction: Transaction): void {
+    this.removeState(transaction.hash())
+  }
+
+  private fetchFromNextSource(hash: TransactionHash): void {
+    if (this.isResolved(hash)) {
+      this.removeState(hash)
+      return
+    }
+
+    // Clear the previous source
+    const currentState = this.pending.get(hash)
+    currentState && this.cleanupCallbacks(currentState)
+
+    // Get the next source
+    const peer = this.popSource(hash)
+    if (!peer) {
+      this.pending.delete(hash)
+      return
+    }
+
+    const timeout = setTimeout(() => this.sendRequest(hash, peer), 500)
+
+    this.pending.set(hash, {
+      peer,
+      action: 'REQUEST_SCHEDULED',
+      timeout,
+    })
+  }
+
+  private sendRequest(hash: TransactionHash, peer: Peer): void {
+    if (this.isResolved(hash)) {
+      this.removeState(hash)
+      return
+    }
+
+    const message = new PooledTransactionsRequest([hash])
+
+    const sent = peer.send(message)
+
+    if (!sent) {
+      this.fetchFromNextSource(hash)
+      return
+    }
+
+    const onPeerDisconnect = ({ peer, state }: { peer: Peer; state: PeerState }) => {
+      if (state.type === 'DISCONNECTED') {
+        peer.onStateChanged.off(onPeerDisconnect)
+        this.fetchFromNextSource(hash)
+      }
+    }
+    peer.onStateChanged.on(onPeerDisconnect)
+
+    const clearDisconnectHandler = () => {
+      peer.onStateChanged.off(onPeerDisconnect)
+    }
+
+    const timeout = setTimeout(() => {
+      this.fetchFromNextSource(hash)
+    }, 5000)
+
+    this.pending.set(hash, {
+      peer,
+      action: 'IN_FLIGHT',
+      timeout,
+      clearDisconnectHandler,
+    })
+  }
+
+  private popSource(hash: TransactionHash): Peer | undefined {
+    const sources = this.sources.get(hash)
+
+    if (!sources) {
+      return undefined
+    }
+
+    let nextSource
+    for (const source of sources.values()) {
+      nextSource = source
+      break
+    }
+
+    if (nextSource) {
+      sources.delete(nextSource)
+    }
+
+    return nextSource
+  }
+
+  private removeState(hash: TransactionHash): void {
+    const currentState = this.pending.get(hash)
+
+    currentState && this.cleanupCallbacks(currentState)
+
+    this.pending.delete(hash)
+    this.sources.delete(hash)
+  }
+
+  private cleanupCallbacks(state: TxState) {
+    if (state.action === 'IN_FLIGHT') {
+      clearTimeout(state.timeout)
+      state.clearDisconnectHandler()
+    } else if (state.action === 'REQUEST_SCHEDULED') {
+      clearTimeout(state.timeout)
+    }
+  }
+
+  private isResolved(hash: TransactionHash): boolean {
+    // TODO: Also check transactions recently added to the chain
```

Comment:
Yea LRU cache is what I was thinking. ETH has a cache in the mempool but honestly might make more sense to have the cache in `Blockchain.ts`. If we want to cache transactions every time we call [`onConnectBlock`](https://github.com/iron-fish/ironfish/blob/564cee5817b4bd6967dcf8371c86c6ad74fa6868/ironfish/src/blockchain/blockchain.ts#L103) then I feel like it would make more sense to be in the chain since theoretically we could have a transaction that makes it to the chain but was never seen by the mempool and we'd still want to mark that as resolved